### PR TITLE
Add deposit subtreeID

### LINKED ISSUE #472: Deposit subtree needs an unique identifier
### What's wrong

The deposit subtree has no unique identifier. This has problems:

- The client has no way to know which deposit root emitted from DepositSubTreeReady has been processed or reorged.
- The client has no way to know if a subtree root has been finalized in the Rollup.sol.

Note that we can not use the subtree root as a unique identifier, because it's highly probable to deposit the same state leaves with the same pubkeyIDs, tokens, and balances.

### How can we fix it

- The variable `front` in the SubtreeQueue contract looks like a suitable choice for the unique identifier. It's ever-increasing and not duplicating. Return the number `front` in `dequeue` function.
- Add the unique identifier in event IDepositManager::DepositSubTreeReady.
- Add the unique identifier in event Rollup::DepositsFinalised.

### PR DESCRIPTION
**Pull request by ChihChengLiang**
Fix #472.

Note that #472 missed the enqueue part.

### CODE REVIEW

File: contracts/DepositManager.sol [hunk 1]
```diff
@@ -21,16 +23,24 @@ contract SubtreeQueue {
     uint256 public front = 1;
     uint256 public back = 0;
 
+    event DepositSubTreeReady(uint256 subtreeID, bytes32 subtreeRoot);
+
     function enqueue(bytes32 subtreeRoot) internal {
-        back += 1;
-        queue[back] = subtreeRoot;
```
Comment (ChihChengLiang) [inline contracts/DepositManager.sol:26]: the old way has a redundant SLOAD.

File: contracts/DepositManager.sol [hunk 2]
```diff
@@ -21,16 +23,24 @@ contract SubtreeQueue {
...
+        uint256 subtreeID = back + 1;
+        back = subtreeID;
+        queue[subtreeID] = subtreeRoot;
+        emit DepositSubTreeReady(subtreeID, subtreeRoot);
```
Comment (ChihChengLiang) [inline contracts/DepositManager.sol:32]: We forgot to add DepositSubTreeReady event in the `reenqueue` function. But if we add the event in the `enqueue` function, it comes with some benefits:
- we don't worry about forgetting it anywhere. 
- We don't pass the `subtreeID` and `readySubtree` all the way to the `depositFor`.

File: contracts/test/TestRollup.sol [hunk 3]
```diff
@@ -7,14 +7,16 @@ import { Types } from "../libs/Types.sol";
 import { IDepositManager } from "../DepositManager.sol";
 
 contract MockDepositManager is IDepositManager {
-    event EnqueSubtree(bytes32 subtreeRoot);
```
Comment (ChihChengLiang) [inline contracts/test/TestRollup.sol:10]: I should've noticed this is the sign we forget to add `DepositSubTreeReady`

**Review by vaibhavchellani — APPROVED** [review]
LGTM! Thanks for the quick PR!

**Pull request merged**